vowpalwabbit/gd.h

```diff
@@ -23,40 +24,44 @@ void print_audit_features(vw&, example& ec);
 void save_load_regressor(vw& all, io_buf& model_file, bool read, bool text);
 void save_load_online_state(vw& all, io_buf& model_file, bool read, bool text, GD::gd *g = nullptr);
 
-struct multipredict_info { size_t count; size_t step; polyprediction* pred; regressor* reg; /* & for l1: */ float gravity; };
+struct multipredict_info { size_t count; size_t step; polyprediction* pred; weight_parameters& weights; /* & for l1: */ float gravity; };
 
 inline void vec_add_multipredict(multipredict_info& mp, const float fx, uint64_t fi)
 { if ((-1e-10 < fx) && (fx < 1e-10)) return;
-  weight*w    = mp.reg->weight_vector;
-  uint64_t mask = mp.reg->weight_mask;
+  weight_parameters& w = mp.weights;
+  uint64_t mask = w.mask(); 
   polyprediction* p = mp.pred;
-
+  weight_parameters::iterator iter = w.begin();
   fi &= mask;
   uint64_t top = fi + (uint64_t)((mp.count-1) * mp.step);
+  uint64_t i = 0;
   if (top <= mask)
-  { weight* last = w + top;
-    w += fi;
-    for (; w <= last; w += mp.step, ++p)
-      p->scalar += fx **w;
+  {
+	  i += fi;
+	  iter += fi;
+	  for (; i <= top; i+= mp.step, iter += mp.step, ++p)
+		  p->scalar += fx * (*iter);
   }
   else    // TODO: this could be faster by unrolling into two loops
-    for (size_t c=0; c<mp.count; ++c, fi += (uint64_t)mp.step, ++p)
-    { fi &= mask;
-      p->scalar += fx * w[fi];
+  { iter += fi;
+    for (size_t c=0; c<mp.count; ++c, iter += (uint64_t)mp.step, ++p)
+    { 
+      p->scalar += fx * (*iter);
```

Comment:
There is a bug at this point because iter can run over the end of the array.  I've fixed.